server/src/frontend_api.rs

```diff
@@ -1,3 +1,250 @@
+
+    Ok(Json(FrontendResult {
+        toggles: evaluated_features,
+    }))
+}
+
+fn resolve_frontend_features(
+    client_features: ClientFeatures,
+    context: Context,
+) -> Vec<EvaluatedToggle> {
+    let mut engine = EngineState::default();
+    engine.take_state(client_features.clone());
+
+    client_features
+        .features
+        .into_iter()
+        .map(|toggle| {
+            let variant = engine.get_variant(toggle.name.clone(), &context);
+            EvaluatedToggle {
+                name: toggle.name.clone(),
+                enabled: engine.is_enabled(toggle.name, &context),
+                variant: EvaluatedVariant {
+                    name: variant.name,
+                    enabled: variant.enabled,
+                    payload: variant.payload.map(|succ| Payload {
+                        payload_type: succ.payload_type,
+                        value: succ.value,
+                    }),
+                },
+                impression_data: false,
+            }
+        })
+        .collect()
+}
+
+pub fn configure_frontend_api(cfg: &mut web::ServiceConfig) {
+    cfg.service(get_frontend_features);
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use crate::types::{EdgeProvider, FeaturesProvider, TokenProvider};
+    use actix_web::{
+        http::header::ContentType,
+        test,
+        web::{self, Data},
+        App,
+    };
+    use serde_json::json;
+    use unleash_types::{
+        client_features::{ClientFeature, ClientFeatures, Constraint, Operator, Strategy},
+        frontend::{EvaluatedToggle, EvaluatedVariant, FrontendResult},
+    };
+
+    #[derive(Clone, Default)]
+    struct MockDataSource {
+        features: Option<ClientFeatures>,
+    }
+
+    impl MockDataSource {
+        fn with(self, features: ClientFeatures) -> Self {
+            return MockDataSource {
+                features: Some(features),
+            };
+        }
+    }
+
+    impl FeaturesProvider for MockDataSource {
+        fn get_client_features(&self, _token: crate::types::EdgeToken) -> ClientFeatures {
+            self.features
+                .as_ref()
+                .expect("You need to populate the mock data for your test")
+                .clone()
+        }
+    }
+
+    impl TokenProvider for MockDataSource {
+        fn get_known_tokens(&self) -> Vec<crate::types::EdgeToken> {
+            todo!()
+        }
+
+        fn secret_is_valid(&self, _secret: &str) -> bool {
+            true
+        }
+
+        fn token_details(&self, _secret: String) -> Option<crate::types::EdgeToken> {
+            todo!()
+        }
+    }
+
+    impl EdgeProvider for MockDataSource {}
+
+    impl Into<Data<dyn EdgeProvider>> for MockDataSource {
+        fn into(self) -> Data<dyn EdgeProvider> {
+            let client_provider_arc: Arc<dyn EdgeProvider> = Arc::new(self.clone());
+            let provider = Data::from(client_provider_arc);
+            provider
```

Comment:
Could this also be something like `Data::from(client_provider_arc)` and return it right away instead of allocating to the variable `provider`?

Reply:
Yeah, that's left over from me fighting with Actix to get it to respect the underlying type. Updated